[Clang][SME2] Add multi-vector add/sub builtins

kmclaughlin-arm opened the pull request:
Adds the following SME2 builtins:
 - sv(add|sub)
 - sv(add|sub)_za32/za64,
 - sv(add|sub)_write_za32/za64

Other changes in this patch:
 - CGBuiltin.cpp: The GetAArch64SMEProcessedOperands function is created
    to avoid duplicating existing code from EmitAArch64SVEBuiltinExpr.
 - arm_sve.td: The add/sub SME2 builtins which do not operate on ZA have
    been added to arm_sve.td, matching the corrosponding LLVM IR intrinsic
    names which start with @llvm.aarch64.sve for this reason.
 - SveEmitter.cpp: Adds the createCoreHeaderIntrinsics function to remove
    duplicated code in createHeader & createSMEHeader. Uses a new enum
    (ACLEKind) to choose either "__builtin_sme_" or "__builtin_sve_" when
    emitting the intrinsics.

See https://github.com/ARM-software/acle/pull/217/files

kmclaughlin-arm committed:
[Clang][SME2] Add multi-vector add/sub builtins

Adds the following SME2 builtins:
 - sv(add|sub)
 - sv(add|sub)_za32/za64,
 - sv(add|sub)_write_za32/za64

Other changes in this patch:
 - SveEmitter.cpp: Adds the createCoreHeaderIntrinsics function to remove
   duplicated code in createHeader & createSMEHeader. Uses a new enum
   (ACLEKind) to choose either "__builtin_sme_" or "__builtin_sve_"
   when emitting the intrinsics.
 - CGBuiltin.cpp: The GetAArch64SMEProcessedOperands function is created
   to avoid duplicating existing code from EmitAArch64SVEBuiltinExpr.
 - arm_sve.td: The add/sub SME2 builtins which do not operate on ZA have
   been added to arm_sve.td, matching the corrosponding LLVM IR intrinsic
   names which start with @llvm.aarch64.sve for this reason.

See https://github.com/ARM-software/acle/pull/217/files

david-arm reviewed (commented):
I've not done an exhaustive review, but thought I'd leave the comments I have so far!

kmclaughlin-arm committed:
- Rename GetAArch64SMEProcessedOperands to GetAArch64SVEProcessedOperands
- Removed code checking for ICEArguments from EmitAArch64SMEBuiltinExpr
- Added comment over check for IsTupleGetOrSet
- Added the code to print SME attributes back into emitIntrinsic

kmclaughlin-arm commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -9893,24 +9888,37 @@ Value *CodeGenFunction::FormSVEBuiltinResult(Value *Call) {
 
-Value *CodeGenFunction::EmitAArch64SVEBuiltinExpr(unsigned BuiltinID,
-                                                  const CallExpr *E) {
+void CodeGenFunction::GetAArch64SMEProcessedOperands(
```
I think this makes sense, I've changed it to GetAArch64SVEProcessedOperands

kmclaughlin-arm commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -9893,24 +9888,37 @@ Value *CodeGenFunction::FormSVEBuiltinResult(Value *Call) {
 
-  llvm::SmallVector<Value *, 4> Ops;
   for (unsigned i = 0, e = E->getNumArgs(); i != e; i++) {
-    if ((ICEArguments & (1 << i)) == 0)
```
Added the following comment:
```
// Tuple set/get only requires one insert/extract vector, which is
// created by EmitSVETupleSetOrGet.
```

kmclaughlin-arm commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -10272,29 +10291,13 @@ Value *CodeGenFunction::EmitAArch64SMEBuiltinExpr(unsigned BuiltinID,
   getContext().GetBuiltinType(BuiltinID, Error, &ICEArguments);
```
I've removed this, it wasn't needed here now that it's checked in GetAArch64SVEProcessedOperands.

kmclaughlin-arm commented on clang/utils/TableGen/SveEmitter.cpp:
```diff
@@ -1016,29 +1021,24 @@ std::string Intrinsic::mangleName(ClassKind LocalCK) const {
-     << (SMEAttrs.empty() ? "__builtin_sve_" : "__builtin_sme_")
-     << FullName << ")";
-  if (!SMEAttrs.empty())
-    OS << SMEAttrs;
```
The attributes aren't currently handled elsewhere in Clang so this shouldn't have been removed. I've added most of this code back in, but kept the ACLEKind switch in below as this is needed when the builtin kind is SVE but it has SME attributes.

kmclaughlin-arm reviewed (commented):
Thank you for reviewing this @david-arm!

david-arm commented on clang/test/CodeGen/aarch64-sme2-intrinsics/acle_sme2_add.c:
```diff
@@ -0,0 +1,1226 @@
+// CPP-CHECK-NEXT:    ret void
+//
+void test_svadd_za32_vg1x2_f32(uint32_t slice_base, svfloat32x2_t zn) {
+  SVE_ACLE_FUNC(svadd_za32,,_f32,,_vg1x2)(slice_base + 6, zn);
```
Perhaps it would be better to add 7 here instead, since that's the limit of the instruction? Having said that, since the LLVM intrinsics don't accept an immediate offset anyway maybe there isn't any value in adding immediate values for any of the add/sub tests?

david-arm reviewed (commented):
This looks a lot better now @kmclaughlin-arm - thanks for the changes! I just have a couple of comments about the tests that I missed previously...

kmclaughlin-arm committed:
- Removed the add of an immediate offset from the tests
- Added missing tests to acle_sme2_sub.c

kmclaughlin-arm committed:
- Removed incorrect tests added for multi-vector sub

david-arm reviewed (approved):
LGTM! Eccelente! Thanks for the changes @kmclaughlin-arm.

sdesmalen-arm commented on clang/include/clang/Basic/arm_sme.td:
```diff
@@ -263,3 +263,38 @@ multiclass ZAFPOuterProd<string n_suffix> {
+////////////////////////////////////////////////////////////////////////////////
+// SME2 - ADD, SUB
+
+multiclass ZAAddSub<string n_suffix> {
```
nit:  
```suggestion
multiclass AddSubZA<string n_suffix> {
```

kmclaughlin-arm replied:
Hi @sdesmalen-arm, I'm happy to change this, though the reason I chose this name was to match other multiclasses in the file such as ZAStore, ZAAdd, ZAFPOuterProd, etc.

sdesmalen-arm commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -9571,22 +9571,17 @@ Value *CodeGenFunction::EmitSVEStructStore(const SVETypeFlags &TypeFlags,
 
   // Does the store have an offset?
-  if (Ops.size() > 3)
+  if (Ops.size() > (2 + N))
```
`EmitSVEStructStore` does not handle any more intrinsics. Should this be part of a different patch?

kmclaughlin-arm replied:
This change was intended to be part of this patch. Now that EmitSVEStructLoad is being called after GetAArch64SVEProcessedOperands, the first operand will have been split into a series of vector extracts and the number of operands expected here will change depending on the value of N.

sdesmalen-arm commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -10266,35 +10288,13 @@ Value *CodeGenFunction::EmitAArch64SVEBuiltinExpr(unsigned BuiltinID,
 
 Value *CodeGenFunction::EmitAArch64SMEBuiltinExpr(unsigned BuiltinID,
                                                   const CallExpr *E) {
-  // Find out if any arguments are required to be integer constant expressions.
```
The changes to EmitAArch64SMEBuiltinExpr seem like a (significant) non-functional change. Can you separate it out as such so that it can be committed separately?

kmclaughlin-arm committed:
- Run update_cc_test_checks script on add/sub tests

sdesmalen-arm commented on clang/utils/TableGen/SveEmitter.cpp:
```diff
@@ -354,6 +356,9 @@ class SVEEmitter {
   void createHeader(raw_ostream &o);
 
+  // Emits core intrinsics in both arm_sme.h and arm_sve.h
+  void createCoreHeaderIntrinsics(raw_ostream &o, SVEEmitter &Emitter, ACLEKind Kind);
```
80char limit. Can you clang-format this patch?

sdesmalen-arm commented on clang/utils/TableGen/SveEmitter.cpp:
```diff
@@ -1180,6 +1194,32 @@ void SVEEmitter::createIntrinsic(
+      Defs.begin(), Defs.end(), [](const std::unique_ptr<Intrinsic> &A,
+                                   const std::unique_ptr<Intrinsic> &B) {
+        auto ToTuple = [](const std::unique_ptr<Intrinsic> &I) {
+          return std::make_tuple(I->getGuard(), (unsigned)I->getClassKind(), I->getName());
```
80char formatting.

sdesmalen-arm commented on clang/include/clang/Basic/arm_sve.td:
```diff
@@ -1980,3 +1980,12 @@ def SVCNTP_COUNT : SInst<"svcntp_{d}", "n}i", "QcQsQiQl", MergeNone, "aarch64_sv
+////////////////////////////////////////////////////////////////////////////////
+// SME2
+
+let TargetGuard = "sme2" in {
```
Can you add a comment here describing why we add SME2 intrinsics to the arm_sve.td file?

kmclaughlin-arm committed:
- Add __arm_streaming and __arm_shared_za attributes to the tests

kmclaughlin-arm committed:
- Moved the multi-single vector adds to a separate test file
- Ran clang-format to fix formatting issues in SveEmitter.cpp
- Added a comment to arm_sve.td to explain why some SME2 intrinsics are included

kmclaughlin-arm commented on clang/include/clang/Basic/arm_sve_sme_incl.td:
```diff
@@ -257,7 +257,7 @@ class ImmCheck<int arg, ImmCheckType kind, int eltSizeArg = -1> {
 
 class Inst<string n, string p, string t, MergeType mt, string i,
-           list<FlagType> ft, list<ImmCheck> ch, MemEltType met> {
+           list<FlagType> ft, list<ImmCheck> ch, MemEltType met = MemEltTyDefault> {
```
Hi @dtemirbulatov, this change is just to set the default MemEltType to MemEltTyDefault, so that we don't have to provide this in every builtin definition.
